File: scripts/tasks/projects.ts
import { hasEnv, isGitRepo, logStep, runOrThrow } from "../common";
import { MANAGED_PROJECTS } from "../projects.config";

const optional = hasEnv("OPTIONAL");

function requireProjectRepo(name: string, path: string): void {
  if (isGitRepo(path)) {
    return;
  }

  const message = `missing: ${name} (${path})`;
  if (optional) {
    console.log(`skip: ${message}`);
    return;
  }

  throw new Error(message);
}

export function listProjects(): void {
  logStep("Managed projects");
  if (MANAGED_PROJECTS.length === 0) {
    console.log("(none configured)");
    return;
  }

  for (const project of MANAGED_PROJECTS) {
    console.log(`${project.name}: ${project.path}`);
  }
}

export function installProjects(): void {
  logStep("Install managed project dependencies");
  if (MANAGED_PROJECTS.length === 0) {
    console.log("(none configured)");
    return;
  }

  for (const project of MANAGED_PROJECTS) {
    if (!isGitRepo(project.path)) {
      requireProjectRepo(project.name, project.path);
      continue;
    }

    console.log(`project: ${project.name}`);
    runOrThrow(project.installCommand, { cwd: project.path });
  }
}

export function verifyProjects(): void {
  logStep("Run managed project verification");
  if (MANAGED_PROJECTS.length === 0) {
    console.log("(none configured)");
    return;
  }

  for (const project of MANAGED_PROJECTS) {
    if (!isGitRepo(project.path)) {
      requireProjectRepo(project.name, project.path);
      continue;
    }

    console.log(`project: ${project.name}`);
    for (const command of project.verifyCommands) {
      runOrThrow(command, { cwd: project.path });
    }
  }
}

export function doctorProjects(): void {
  logStep("Managed project health");
  if (MANAGED_PROJECTS.length === 0) {
    console.log("(none configured)");
    return;
  }

  for (const project of MANAGED_PROJECTS) {
    const present = isGitRepo(project.path);
    console.log(
      `${project.name}: ${present ? "ok" : "missing"} (${project.path})`,
    );
    if (!present) {
      continue;
    }

    const branch = runOrThrow(["git", "branch", "--show-current"], {
      cwd: project.path,
      quiet: true,
    });

    const pushUrls = runOrThrow(
      ["git", "remote", "get-url", "--all", "--push", "origin"],
      {
        cwd: project.path,
        quiet: true,
      },
    );

    const urls = pushUrls
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
      .join(" | ");

    console.log(`branch: ${branch}`);
    console.log(`push: ${urls}`);
  }
}

File: scripts/tasks/projects.py
"""Managed project operations: list, install, verify, doctor."""

from __future__ import annotations

from scripts.common import (
    git_remote_push_urls,
    has_env,
    is_git_repo,
    log_step,
    run_or_throw,
)
from scripts.projects_config import MANAGED_PROJECTS


def _optional_mode() -> bool:
    return has_env("OPTIONAL")


def _require_project_repo(name: str, path: str) -> None:
    if is_git_repo(path):
        return

    message = f"missing: {name} ({path})"
    if _optional_mode():
        print(f"skip: {message}")
        return

    raise RuntimeError(message)


def list_projects() -> None:
    log_step("Managed projects")
    if not MANAGED_PROJECTS:
        print("(none configured)")
        return

    for project in MANAGED_PROJECTS:
        print(f"{project.name}: {project.path}")


def install_projects() -> None:
    log_step("Install managed project dependencies")
    if not MANAGED_PROJECTS:
        print("(none configured)")
        return

    for project in MANAGED_PROJECTS:
        path = str(project.path)
        if not is_git_repo(path):
            _require_project_repo(project.name, path)
            continue

        print(f"project: {project.name}")
        run_or_throw(project.install_command, cwd=path)


def verify_projects() -> None:
    log_step("Run managed project verification")
    if not MANAGED_PROJECTS:
        print("(none configured)")
        return

    for project in MANAGED_PROJECTS:
        path = str(project.path)
        if not is_git_repo(path):
            _require_project_repo(project.name, path)
            continue

        print(f"project: {project.name}")
        for command in project.verify_commands:
            run_or_throw(command, cwd=path)


def doctor_projects() -> None:
    log_step("Managed project health")
    if not MANAGED_PROJECTS:
        print("(none configured)")
        return

    for project in MANAGED_PROJECTS:
        path = str(project.path)
        present = is_git_repo(path)
        status = "ok" if present else "missing"
        print(f"{project.name}: {status} ({path})")
        if not present:
            continue

        branch = run_or_throw(["git", "branch", "--show-current"], cwd=path, quiet=True)

        print(f"branch: {branch}")

        origin_urls = git_remote_push_urls(path, "origin")
        if origin_urls:
            print(f"push(origin): {' | '.join(origin_urls)}")

        fork_urls = git_remote_push_urls(path, "fork")
        if fork_urls:
            print(f"push(fork): {' | '.join(fork_urls)}")
